### PR TITLE
Translated date times via standard Unicode formats.

### PR DESCRIPTION
**Pull request by Daniel-KM**
The module doesn't render formats in other language than English, so I added an option for that. The same option can be added in omeka core (see pull request).

### CODE REVIEW

**Comment by jimsafley** [comment]
Thanks. I can't find the PR you reference, but obviously I can only merge this PR if/when that one's merged.

**Comment by Daniel-KM** [comment]
In fact, there was no pr in omeka, because since 3.1, there is a new option for lang, so I updated the code here. In fact, here this is not the language of the value, but the language of the interface, so it is used by default, and can be overridden. 

I added a pr on omeka/omeka-s#1802 to add the option in the datatype interface for render (similar arguments as rendering medias), but because it is used now for lang, it's a little more complex to manage.

**Comment by jimsafley** [comment]
I agree that the rendered dates should be localized. `IntlDateFormatter` is the obvious solution, but it comes with a few complications, as you already know:

- Ideally we'd set a specific `$dateType` and `$timeType` and let `IntlDateFormatter` calculate the default pattern for the locale. This would give us localized format and translation for free, but, for some reason, the formatted date does include an era (e.g. BC and AD) or otherwise identify a year as being negative. So we are forced to set our own patterns.
- As you note, the representation of negative dates is tricky, perhaps insurmountable, due to the leap year problem.
- We must explicitly set the timezone on `IntlDateFormatter` or the formatted date will normalize the time to the server's timezone and set the formatted offest to `00:00`. For example `08:00:00 +02:00` becomes ` 06:00:00 +00:00`. I much prefer that we retain the original offset on render.

I really wish we could find a way to simply use `$dateType` and `$timeType` so we wouldn't have to bother with format patterns. But, even so, unless we can find some way to consistently and accurately represent negative dates, I'm hesitant to move forward on this PR, which looks to be a work in progress anyway.

Maybe we can try another approach, like allow administrators to configure the patterns themselves, and somehow interpolate the translated months, eras, etc. at runtime. Not sure if it's workable, but it's an idea.

**Comment by jimsafley** [comment]
We also have to address the possibility that [`IntlDateFormatter`](https://www.php.net/manual/en/class.intldateformatter.php) (and likely the upcoming [`IntlDatePatternGenerator`](https://wiki.php.net/rfc/intldatetimepatterngenerator)) just won't work for the NumericDataTypes module. Take the following examples, noting the inconsistencies between the set date and the formatted date:

```php
$dateFormatter = new IntlDateFormatter(null, IntlDateFormatter::NONE, IntlDateFormatter::NONE);
$dateFormatter->setPattern('d LLLL yyyy G');
$dateTime = new DateTime;

$dateTime->setDate(-1, 1, 1);
echo $dateFormatter->format($dateTime);
// 3 January 0002 BC

$dateTime->setDate(-1000000, 1, 1);
echo $dateFormatter->format($dateTime);
// 16 July 999981 BC

php > $dateTime->setDate(1, 1, 1);
php > echo $dateFormatter->format($dateTime);
// 3 January 0001 AD

$dateTime->setDate(1000, 1, 1);
echo $dateFormatter->format($dateTime);
// 27 December 0999 AD

$dateTime->setDate(100000, 1, 1);
echo $dateFormatter->format($dateTime);
// 1 January 100000 AD
```

I think the inconsistency we're seeing here is that `IntlDateFormatter` automatically adjusts for a) The Gregorian switch, b) leap year, and c) the fact that there was no year zero.

Regardless, these adjustments are not compatible with how the module processes and renders dates. The module formats dates as-is—as they were entered by the user. So $dateTime->setDate(-1000000, 1, 1)  should render "1 January -1000000 " and $dateTime->setDate(1000, 1, 1) should render "1 January 1000". I don't see a way for `IntlDateFormatter` to do this.

Put simply, I'm not sure how we're going to reconcile the inconsistency between the dates the user enters and the dates that the module renders to screen when using `IntlDateFormatter`.

**Comment by Daniel-KM** [comment]
Yes, IntlDateFormatter is not the good way. Or it should be limited to dates after 1582, but it depends on countries (us/uk starts using Gregorian calendar in 1752). 

The point is the `$date['date']->format($date['format_render'])`, there should be a way to config the `format_render` and to set it by default to the language of the site, but they are currently hard coded. 

So the simplest way is just to add "@translate" to it and to translate the format before passing it to format() as any translation. And for user who want to use a specific format, to use the option.

**Comment by Daniel-KM** [comment]
I updated the module for a new way that keep the original date, and translate directly days and monthes. Not perfect, but working enough.

**Comment by Daniel-KM** [comment]
You can remove the update of the language, but it is needed to check if the module is working fine.

**Comment by jimsafley** [comment]
We're currently working on an alternative solution that draws inspiration from this PR. The [intl-date-formatter](https://github.com/omeka-s-modules/NumericDataTypes/compare/intl-date-formatter) branch rectifies some of my concerns about using `IntlDateFormatter` and it's compatible with a future inclusion of [`IntlDatePatternGenerator`](https://wiki.php.net/rfc/intldatetimepatterngenerator).
